ArcGISRuntimeSDKQt_CppSamples/Scenes/OpenMobileScenePackage/OpenMobileScenePackage.cpp

```diff
@@ -0,0 +1,153 @@
+#include <QTemporaryDir>
+#include <QtCore/qglobal.h>
+
+#ifdef Q_OS_IOS
+#include <QStandardPaths>
+#endif // Q_OS_IOS
+
+using namespace Esri::ArcGISRuntime;
+
+OpenMobileScenePackage::OpenMobileScenePackage(QObject* parent /* = nullptr */):
+  QObject(parent)
+{  
+}
+
+OpenMobileScenePackage::~OpenMobileScenePackage() = default;
+
+// helper method to get cross platform data path
+namespace {
+  QString defaultDataPath()
+  {
+    QString dataPath;
+
+  #ifdef Q_OS_ANDROID
+    dataPath = "/sdcard";
+  #elif defined Q_OS_IOS
+    dataPath = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
+  #else
+    dataPath = QDir::homePath();
+  #endif
+
+    return dataPath;
+  }
+}
+
+void OpenMobileScenePackage::init()
+{
+  // Register classes for QML
+  qmlRegisterType<SceneQuickView>("Esri.Samples", 1, 0, "SceneView");
+  qmlRegisterType<OpenMobileScenePackage>("Esri.Samples", 1, 0, "OpenMobileScenePackageSample");
+}
+
+SceneQuickView* OpenMobileScenePackage::sceneView() const
+{
+  return m_sceneView;
+}
+
+// Set the view (created in QML)
+void OpenMobileScenePackage::setSceneView(SceneQuickView* sceneView)
+{
+  if (!sceneView || sceneView == m_sceneView)
+  {
+    return;
+  }
+
+  m_sceneView = sceneView;
+
+  openPackage();
+
+  emit sceneViewChanged();
+}
+
+// Open a mobile scene package
+void OpenMobileScenePackage::openPackage()
+{
+  // create the MSPK data path
+  const QString dataPath = defaultDataPath() + "/ArcGIS/Runtime/Data/mspk/philadelphia.mspk";
+
+  // connect to the Mobile Scene Package to determine if direct read is supported
+  connect(MobileScenePackage::instance(), &MobileScenePackage::isDirectReadSupportedCompleted,
+          this, [this, dataPath](QUuid, bool supported)
+  {
+    // if direct read is supported, load the package
+    if (supported)
+    {
+      createScenePackage(dataPath);
+    }
+    // otherwise, the package needs to be unpacked
+    else
+    {
+      m_unpackTempPath = QTemporaryDir().path();
+      MobileScenePackage::unpack(dataPath, m_unpackTempPath);
+    }
+  });
+
+  // connect to the Mobile Scene Package to know when the data is unpacked
+  connect(MobileScenePackage::instance(), &MobileScenePackage::unpackCompleted,
+          this, [this](QUuid, bool success)
+  {
+    // if the unpack was successful, load the unpacked package
+    if (success)
+    {
+      createScenePackage(m_unpackTempPath);
+    }
+    // log that the upack failed
+    else
+    {
+      qDebug() << "failed to unpack";
+    }
+  });
+
+  // Check if the MSPK can be read directly
+  MobileScenePackage::isDirectReadSupported(dataPath);
+}
+
+// Slot for handling when the package loads
+void OpenMobileScenePackage::packageLoaded(Error e)
+{
+  if (!e.isEmpty())
+  {
+    qDebug() << QString("Package load error: %1 %2").arg(e.message(), e.additionalMessage());
+    return;
+  }
+
+  if (m_scenePackage->scenes().length() < 1)
+    return;
+
+  // obtain the first scene in the list of scenes
```

Comment:
add a comment to explain that the package could have multiple scenes to choose from - and that the user could be prompted to do that in UI?